docs/configuration.md
# Configuration File

By default the network importer will try to load a configuration file name `network_importer.toml`, this configuration file is not mandatory as long as the required information to connect to the SOT, Batfish and/or the network devices are provided via environment variables.

It's possible to specify which configuration file should be used in cli with the option `--config`.

The configuration file is organized in 7 sections detailed below. 

## Main Section

The main section included the main parameters to define what will be imported from the network and/or the SOT, where the configuration files are available/should be stored and how many number of tasks can be executed in parallel.

```toml
[main]
import_ips = true 
import_prefixes = false
import_intf_status = false

# Vlans (name, vlan-id) can be imported from the configuration, from the CLI or both 
# - "config" will import the vlans just from the configuration
# - "cli" will import the vlans from the cli using the action `get_vlans`
# - "true" will import vlans from both cli and config
# - no or false will not import any vlans
# The association between interface and vlans will always be derived from the configuration.
import_vlans = "config"         # Valid options are ["cli", "config", "no", true, false]

# Cabling can be imported from LLDP, CDP or the configuration (for some point to point links)
# - "lldp" or "cdp" will import the vlans from the cli using the action `get_neighbors`
# - "config" will import the neighbors from the configuration (for point to point links)
# - "true" will import neighbors from both cli and config 
# - no or false will not import any neighbors
import_cabling = "lldp"         # Valid options are ["lldp", "cdp", "config", "no", true, false]
excluded_platforms_cabling = ["cisco_asa"]

# Number of Nornir tasks to execute at the same time
nbr_workers = 25

# Directory where the configuration can be find, organized in Batfish format
configs_directory = "configs"

# Valid Backend
# Only Netbox and Nautobot backend are included by default, if you want to use another backend
# you must leave backend empty and define inventory.inventory_class and adapters.sot_class manually.
backend = "nautobot"
```

# Inventory Section

The `[inventory]` section regroup all parameters related to the inventory and also includes an optional list of supported platforms. network-importer supports multiple inventories, you can define your own inventory by defining `inventory_class` and each inventory can define its own list of settings which will need to be configured under `[inventory.settings]`. The settings for the [nautobot inventory](backend/nautobot.md) and for the [netbox inventory](backend/netbox.md) are available in their respective documentation.

```toml
[inventory]

# Define a list of supported platform, 
# if defined all devices without platform or with a different platforms will be removed from the inventory
supported_platforms = [ "cisco_ios", "cisco_nxos" ]

# Configure which Inventory will be loaded by the network importer.
backend = "nautobot"

[inventory.settings]
# Inventory specific settings, please refer to the documentation of each backend/inventory.
```

## Batfish Section

The `[batfish]` section regroup all parameters to connect to Batfish.
```toml
[batfish]
address = "localhost"               # Alternative Env Variable : BATFISH_ADDRESS
api_key = "XXXX"                    # Alternative Env Variable : BATFISH_API_KEY
network_name = "network-importer"   # Alternative Env Variable : BATFISH_NETWORK_NAME
snapshot_name = "latest"            # Alternative Env Variable : BATFISH_SNAPSHOT_NAME
port_v1 = 9997                      # Alternative Env Variable : BATFISH_PORT_V1
port_v2 = 9996                      # Alternative Env Variable : BATFISH_PORT_V2
use_ssl = false                     # Alternative Env Variable : BATFISH_USE_SSL
```

## Network Section

To be able to pull live information from the devices, the credential information needs to be provided either in the configuration file or as environment variables.

It's also possible to define some connection parameters for Netmiko and define a list of expected FDQNs that can be found in the network.

```toml
[network]
login = "username"          # Alternative Env Variable : NETWORK_DEVICE_LOGIN
password = "password"       # Alternative Env Variable : NETWORK_DEVICE_PWD
enable = true               # Alternative Env Variable : NETWORK_DEVICE_ENABLE

# List of valid FQDN that can be found in the network,
# The FQDNs in this list will be automatically removed from all neighbors discovered from LLDP/CDP
fqdns = [ ]

[network.netmiko_extras]
# Any additional parameters for Netmiko defined in this section will be automatically configured 
# as part of the Nornir inventory.

[network.napalm_extras]
# Any additional parameters for Napalm defined in this section will be automatically configured 
# as part of the Nornir inventory.
```

<!-- ## Adapters Section

Configure which adapters will be loaded by the network importer.
Please see the [extensibility section](extensibility.md) of the documentation for more details on how to create your own adapter.
The settings for the [nautobot adapter](backend/nautobot.md) and for the [netbox adapter](backend/netbox.md) are available in their respective documentation.

```toml
[adapters]
network_class = "network_importer.adapters.network_importer.adapter.NetworkImporterAdapter"


sot_class = "network_importer.adapters.netbox_api.adapter.NetBoxAPIAdapter"

[adapters.network_settings]
# Network Adapter specific settings, any settings defined in this section will be passed to the Network Adapter. 
# The default network adapter doesn't accept additional settings, nonetheless this section remains available if you want to
# customize the default Network Adapter

[adapters.sot_settings]
# SOT Adapter specific settings, any settings defined in this section will be passed to the SOT Adapter. 
# Please refer to the documentation of each adapter to see what settings are required/supported.
``` -->

## Drivers Section

Configure which driver to use for a given platform.
Please see the [extensibility section](extensibility.md) of the documentation for more details on how to create your own driver.

```toml
[drivers.mapping]
default = "network_importer.drivers.default"
cisco_nxos = "network_importer.drivers.cisco_default"
cisco_ios = "network_importer.drivers.cisco_default"
cisco_xr = "network_importer.drivers.cisco_default"
juniper_junos = "network_importer.drivers.juniper_junos"
arista_eos = "network_importer.drivers.arista_eos"
```

## Logs Section

Control how the application is generating logs.

```toml
[logs]
# Control the level of the logs printed to the console.
level = "info"        # "debug", "info", "warning"

# For each run, a performance log can be generated to capture how long
# some functions took to execute
performance_log = false
performance_log_directory = "performance_logs"
```


## Environment variables

| Environment Variable  | config section         | config file entry          |
| --------------------- | ---------------------- | -------------------------- |
|                       | main                   | backend                    |
|                       | main                   | import_ips                 |
|                       | main                   | import_prefixes            |
|                       | main                   | import_intf_status         |
|                       | main                   | import_vlans               |
|                       | main                   | import_cabling             |
|                       | main                   | excluded_platforms_cabling | 
|                       | main                   | nbr_workers                |
|                       | main                   | configs_directory          |
| BATFISH_ADDRESS       | batfish                | address                    |
| BATFISH_NETWORK_NAME  | batfish                | network_name               |
| BATFISH_SNAPSHOT_NAME | batfish                | snapshot_name              |
| BATFISH_API_KEY       | batfish                | api_key                    |
| BATFISH_PORT_V1       | batfish                | port_v1                    |
| BATFISH_PORT_V2       | batfish                | port_v2                    |
| BATFISH_USE_SSL       | batfish                | use_ssl                    |
|                       | drivers.mapping        | <should this be made into ENV variables?> |
|                       | inventory              | supported_platforms        |
|                       | inventory              | backend (duplicate of main.backend?) |
|                       | inventory.settings     | <TODO>                     |
|                       | logs                   | level                      |
|                       | logs                   | performance_log            |
|                       | logs                   | performance_log_directory  |
| NETWORK_DEVICE_LOGIN  | network                | login                      |
| NETWORK_DEVICE_PWD    | network                | password                   |
| NETWORK_DEVICE_ENABLE | network                | enable                     |
|                       | network                | fqdns                      |
|                       | network.netmiko_extras | <TODO>                     |
|                       | network.napalm_extras  | <TODO>                     |


network_importer/config.py
"""Settings definition for the network importer."""
# pylint: disable=invalid-name,redefined-outer-name

import os
import sys
from pathlib import Path
from typing import List, Dict, Optional, Union

import toml
from pydantic import BaseSettings, ValidationError
from typing_extensions import Literal

from network_importer.exceptions import ConfigLoadFatalError

SETTINGS = None

DEFAULT_DRIVERS_MAPPING = {
    "default": "network_importer.drivers.default",
    "cisco_nxos": "network_importer.drivers.cisco_default",
    "cisco_ios": "network_importer.drivers.cisco_default",
    "cisco_xr": "network_importer.drivers.cisco_default",
    "juniper_junos": "network_importer.drivers.juniper_junos",
    "arista_eos": "network_importer.drivers.arista_eos",
}

DEFAULT_BACKENDS = {
    "netbox": {
        "inventory": "NetBoxAPIInventory",
        "adapter": "network_importer.adapters.netbox_api.adapter.NetBoxAPIAdapter",
    },
    "nautobot": {
        "inventory": "NautobotAPIInventory",
        "adapter": "network_importer.adapters.nautobot_api.adapter.NautobotAPIAdapter",
    },
}

# pylint: disable=global-statement


class BatfishSettings(BaseSettings):
    """Settings definition for the Batfish section of the configuration."""

    address: str = "localhost"
    network_name: str = "network-importer"
    snapshot_name: str = "latest"
    port_v1: int = 9997
    port_v2: int = 9996
    use_ssl: bool = False
    api_key: Optional[str]

    class Config:
        """Additional parameters to automatically map environment variable to some settings."""

        fields = {
            "address": {"env": "BATFISH_ADDRESS"},
            "network_name": {"env": "BATFISH_NETWORK_NAME"},
            "snapshot_name": {"env": "BATFISH_SNAPSHOT_NAME"},
            "api_key": {"env": "BATFISH_API_KEY"},
            "port_v1": {"env": "BATFISH_PORT_V1"},
            "port_v2": {"env": "BATFISH_PORT_V2"},
            "use_ssl": {"env": "BATFISH_USE_SSL"},
        }


class NetworkSettings(BaseSettings):
    """Settings definition for the Network section of the configuration."""

    login: Optional[str]
    password: Optional[str]
    enable: bool = True

    netmiko_extras: Optional[dict]
    napalm_extras: Optional[dict]

    fqdns: List[str] = list()  # List of valid FQDN that can be found in the network

    class Config:
        """Additional parameters to automatically map environment variable to some settings."""

        fields = {
            "login": {"env": "NETWORK_DEVICE_LOGIN"},
            "password": {"env": "NETWORK_DEVICE_PWD"},
            "enable": {"env": "NETWORK_DEVICE_ENABLE"},
        }


class LogsSettings(BaseSettings):
    """Settings definition for the Log section of the configuration."""

    level: Literal["debug", "info", "warning"] = "info"
    # directory: str = "logs"
    performance_log: bool = False
    performance_log_directory: str = "performance_logs"
    # change_log: bool = True
    # change_log_format: Literal[
    #     "jsonlines", "text"
    # ] = "text"  # dict(type="string", enum=["jsonlines", "text"], default="text"),
    # change_log_filename: str = "changelog"


class MainSettings(BaseSettings):
    """Settings definition for the Main section of the configuration."""

    import_ips: bool = True
    import_prefixes: bool = False
    import_cabling: Union[bool, Literal["lldp", "cdp", "config", "no"]] = "lldp"
    excluded_platforms_cabling: List[str] = list()

    import_vlans: Union[bool, Literal["config", "cli", "no"]] = "config"
    import_intf_status: bool = False

    nbr_workers: int = 25

    configs_directory: str = "configs"

    backend: Optional[Literal["nautobot", "netbox"]]
    """Only Netbox and Nautobot backend are included by default, if you want to use another backend
    you must leave backend empty and define inventory.inventory_class and adapters.sot_class manually."""

    # NOT SUPPORTED CURRENTLY
    generate_hostvars: bool = False
    hostvars_directory: str = "host_vars"


class AdaptersSettings(BaseSettings):
    """Settings definition for the Adapters section of the configuration."""

    network_class: str = "network_importer.adapters.network_importer.adapter.NetworkImporterAdapter"
    network_settings: Optional[dict]

    sot_class: Optional[str]
    sot_settings: Optional[dict]


class DriversSettings(BaseSettings):
    """Settings definition for the Drivers section of the configuration."""

    mapping: Dict[str, str] = DEFAULT_DRIVERS_MAPPING


class InventorySettings(BaseSettings):
    """Settings definition for the Inventory section of the configuration.

    By default, the inventory will use the primary IP to reach out to the devices
    if the use_primary_ip flag is disabled, the inventory will try to use the hostname to the device
    """

    inventory_class: Optional[str]
    settings: Optional[dict]

    supported_platforms: List[str] = list()


class Settings(BaseSettings):
    """Main Settings Class for the project.

    The type of each setting is defined using Python annotations
    and is validated when a config file is loaded with Pydantic.
    """

    main: MainSettings = MainSettings()
    batfish: BatfishSettings = BatfishSettings()
    logs: LogsSettings = LogsSettings()
    network: NetworkSettings = NetworkSettings()
    adapters: AdaptersSettings = AdaptersSettings()
    drivers: DriversSettings = DriversSettings()
    inventory: InventorySettings = InventorySettings()


def _configure_backend(settings: Settings):
    """Configure the inventory and the SOT adapter for a given backend.

    The inventory and/or the adapter will be updated if:
      - a backend provided is valid
      - no value has been defined already

    Args:
        settings (Settings)

    Returns:
        Settings
    """
    if not settings.main.backend and (not settings.inventory.inventory_class or not settings.adapters.sot_class):
        raise ConfigLoadFatalError(
            "You must define a valid backend or assign inventory.inventory_class and adapters.sot_class manually."
        )

    if not settings.main.backend:
        return settings

    supported_backends = DEFAULT_BACKENDS.keys()
    if settings.main.backend not in supported_backends:
        raise ConfigLoadFatalError(f"backend value one of : {', '.join(supported_backends)}")

    if not settings.inventory.inventory_class:
        settings.inventory.inventory_class = DEFAULT_BACKENDS[settings.main.backend]["inventory"]

    if not settings.adapters.sot_class:
        settings.adapters.sot_class = DEFAULT_BACKENDS[settings.main.backend]["adapter"]

    return settings


def load(config_file_name="network_importer.toml", config_data=None):
    """Load configuration.

    Configuration is loaded from a file in network_importer.toml format that contains the settings,
    or from a dictionary of those settings passed in as "config_data"

    Args:
        config_file_name (str, optional): Name of the configuration file to load. Defaults to "network_importer.toml".
        config_data (dict, optional): dict to load as the config file instead of reading the file. Defaults to None.
    """
    global SETTINGS

    if config_data:
        SETTINGS = _configure_backend(Settings(**config_data))
        return

    if os.path.exists(config_file_name):
        config_string = Path(config_file_name).read_text()
        config_tmp = toml.loads(config_string)
        SETTINGS = _configure_backend(Settings(**config_tmp))
        return

    SETTINGS = Settings()


def load_and_exit(config_file_name="network_importer.toml", config_data=None):
    """Calls load, but wraps it in a try except block.

    This is done to handle a ValidationError which is raised when settings are specified but invalid.
    In such cases, a message is printed to the screen indicating the settings which don't pass validation.

    Args:
        config_file_name (str, optional): [description]. Defaults to "network_importer.toml".
        config_data (dict, optional): [description]. Defaults to None.
    """
    try:
        load(config_file_name=config_file_name, config_data=config_data)
    except ValidationError as err:
        print(f"Configuration not valid, found {len(err.errors())} error(s)")
        for error in err.errors():
            print(f"  {'/'.join(error['loc'])} | {error['msg']} ({error['type']})")
        sys.exit(1)
    except ConfigLoadFatalError as err:
        print("Configuration not valid")
        print(f"  {err}")
        sys.exit(1)
